File: Qsort.py
#!/usr/bin/env python3

from sys import argv



def quicksort(a, left, right):
    if left < right:
        pivotIdx = left + ((right - left) // 2)
        newPivotIdx = partition(a, left, right, pivotIdx)
        quicksort(a, left, newPivotIdx - 1)
        quicksort(a, newPivotIdx + 1, right)


def partition(a, left, right, pivotIdx):
    pivotVal = a[pivotIdx]
    a[pivotIdx], a[right] = a[right], a[pivotIdx]
    storeIdx = left
    for i in range(left, right):
        if(a[i] < pivotVal):
            a[i], a[storeIdx] = a[storeIdx], a[i]
            storeIdx += 1

    a[storeIdx], a[right] = a[right], a[storeIdx]
    return storeIdx


def init_array(filename):
    with open(filename, 'r') as f:
        for line in f:
            yield int(line)


def print_array(a, filename):
    with open(filename, 'w') as f:
        f.write('\n'.join(str(n) for n in a))


def main():
    if len(argv) != 2:
        print("Usage ./Qsort.py <powerOfTen>")
        return
    IDX = int(argv[1])

    ifName = "10e{}.txt".format(IDX)
    a = list(init_array(ifName))
    quicksort(a, 0, len(a) - 1)
    #Used to test sorting
    # print_array(a, "10e{}sortedPY.txt".format(IDX))
    print("Sorted {} ints in python.".format(len(a)))

if __name__ == '__main__':
    main()

File: Qsort.php
#!/usr/bin/php
<?php

  function quicksort(& $a, $left, $right)
  {
	  if ($left < $right) {
		  $pivotIdx = $left + (($right - $left) / 2);
		  $newPivotIdx = partition($a, $left, $right, $pivotIdx);
		  quicksort($a, $left, $newPivotIdx - 1);
		  quicksort($a, $newPivotIdx + 1, $right);
	}
  }

  function partition(& $a, $left, $right, $pivotIdx)
  {
	  $pivotVal = $a[$pivotIdx];
	  swap($a, $pivotIdx, $right);
	  $storeIdx = $left;
	  for($i = $left; $i < $right; $i++) {
		  if($a[$i] < $pivotVal) {
			  swap($a, $i, $storeIdx);
			  $storeIdx++;
		  }
	}
	  swap($a, $storeIdx, $right);
	  return $storeIdx;
  }

  function swap(& $a, $x, $y)
  {
	  $tmp = $a[$x];
	  $a[$x] = $a[$y];
	  $a[$y] = $tmp;
  }

  function init_array(& $a, $filename, $size)
  {
	  $iF = fopen($filename, "r"); $i = 0;
	  while($i < $size) {
	    $a[$i++] = (int)fgets($iF);
	 }
	  fclose($iF);
  }
/*
  function print_array(& $a, $filename)
  {
	  $oF = fopen($filename, "w");
	  for($i = 0; $i < count($a) - 1; $i++) {
	    fputs($oF, $a[$i]."\n");
	  }
	  fclose($oF);
  }
*/

  
  function pow_ten($idx)
  {
	if($idx == 0) 
		return 1;
	 else 
		 return 10 * pow_ten($idx - 1);
  }

  function main($argc, $argv)
  {
	  if($argc != 2) {
		echo "Usage Qsort.php <power of 10>".PHP_EOL;
		return;
	  }
	  $IDX = (int)$argv[1];
	  $ARRAY_SIZE = pow_ten($IDX);
	  $a = new SplFixedArray($ARRAY_SIZE);
	  $ifName = "10e".$IDX.".txt";
	  //$ofName = "10e".$IDX."sortedPHP.txt";
	  init_array($a, $ifName, $ARRAY_SIZE);
	  quicksort($a, 0, $ARRAY_SIZE - 1);
	  //Used to test sorting
	  //print_array($a, $ofName);
	  unset($a);
	  echo "Sorted ".$ARRAY_SIZE." ints in php".PHP_EOL;
  }
  ini_set('memory_limit','8G');
  main($argc, $argv);
?>
